GVFS/GVFS.Platform.Windows/WindowsVolumeStateWatcher.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Management;
+using GVFS.Common.FileSystem;
+
+namespace GVFS.Platform.Windows
+{
+    public class WindowsVolumeStateWatcher : IVolumeStateWatcher
+    {
+        private readonly ManagementEventWatcher volumeWatcher;
+        private readonly ManagementEventWatcher cryptoVolumeWatcher;
+        private readonly IDictionary<string, bool> cryptoVolumeLockStatuses =
+            new ConcurrentDictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+
+        public WindowsVolumeStateWatcher(TimeSpan pollingInterval)
+        {
+            int intervalSeconds = (int)pollingInterval.TotalSeconds;
+
+            // Watch for mount and unmount of volumes
+            string volumeQuery = $"SELECT * FROM Win32_VolumeChangeEvent WITHIN {intervalSeconds}";
+            this.volumeWatcher = new ManagementEventWatcher(volumeQuery);
+            this.volumeWatcher.EventArrived += this.OnVolumeEvent;
+
+            // Watch for changes to BitLocker-protected volume states (unlock, lock, etc)
+            string cryptoVolumeQuery = $"SELECT * FROM __InstanceModificationEvent WITHIN {intervalSeconds} WHERE TargetInstance ISA 'Win32_EncryptableVolume'";
+            this.cryptoVolumeWatcher = new ManagementEventWatcher(BitLockerHelpers.VolumeEncryptionNamespace, cryptoVolumeQuery);
+            this.cryptoVolumeWatcher.EventArrived += this.OnCryptoVolumeEvent;
+        }
+
+        public event EventHandler<VolumeStateChangedEventArgs> VolumeStateChanged;

```

Comment:
What benefit are we getting here from using a C# event vs just a callback delegate? I don't think we want/need to support multiple listeners, right? Events tend to make diagnostics a little harder, so I would constrain this down if we don't need the flexibility.